self-api/src/main/java/com/selfxdsd/api/ConditionalResource.java

```diff
@@ -0,0 +1,213 @@
+/**
+ * Copyright (c) 2020-2021, Self XDSD Contributors
+ * All rights reserved.
+ * <p>
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"),
+ * to read the Software only. Permission is hereby NOT GRANTED to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software.
+ * <p>
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+ * OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.selfxdsd.api;
+
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import java.net.URI;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Conditional {@link Resource}.
+ * A Resource is considered conditional if it has a eTag header in its
+ * response headers.
+ * @author criske
+ * @version $Id$
+ * @since 0.0.8
+ */
+public interface ConditionalResource extends Resource {
+
+    /**
+     * Uri of this resource. This should be also considered the
+     * storing key.
+     *
+     * @return URI.
+     */
+    URI uri();
+
+    /**
+     * ETag of this resource.
+     *
+     * @return String.
+     */
+    String etag();
+
+    /**
+     * Creation date of the conditional resource.
+     *
+     * @return LocalDateTime.
+     */
+    LocalDateTime creationDate();
+
+    /**
+     * Factory method that creates a ConditionalResource from a regular
+     * Resource.
+     * If the regular can't be conditional
+     * (no ETag header present in its response headers) it returns null.
+     *
+     * @param uri URI associated with the resource.
+     * @param resource Original Resource.
+     * @return ConditionalResource or null.
+     */
+    static ConditionalResource fromResource(
```

Comment:
@criske Please avoid static factory methods. Simply use FromHeaders instead, it should have a constructor which takes Resource

Reply:
@amihaiemil If we're doing this then, when Resource is not cacheable (no etag in headers) we need to throw an exception in constructor instead and therefore making things a little bit ugly in ConditionalResources (we need to catch it). This is why FromResource constructor is private and let this factory to take control and return null otherwise. Also the factory will not create a new CachedResource if the provided Resource is already cached.

However I think it will be cleaner and more intuitive if we're moving the factory to `FromResource` class scope or, even better, moving the whole class FromResource along with its factory to self-impl. I don't think we'll ever use this class outside of self-impl (in self-storage surely not).

Reply:
@criske ok, then let's leave this as it is :D